File: experiments-local/experiment-17/runs/agent=omp_language=typescript_model=openrouter/deepseek/deepseek-v4-pro_tooling=none/rep3/src/team-normalizer.ts
/**
 * Brazilian Soccer MCP Server - Team Name Normalizer
 *
 * Handles team name variations across datasets:
 * - With state suffix: "Palmeiras-SP", "Flamengo-RJ"
 * - Without suffix: "Palmeiras", "Flamengo"
 * - Full names: "Sport Club Corinthians Paulista"
 * - Accented names: "São Paulo", "Grêmio", "Avaí"
 * - Cedilla: "Fortaleza Esporte Clube"
 */

const TEAM_NAME_MAP: Map<string, string> = new Map();

// Initialize known Brazilian team name mappings
const BRAZILIAN_CLUBS: Record<string, string[]> = {
  "Flamengo": ["Flamengo", "Flamengo-RJ", "Clube de Regatas do Flamengo"],
  "Fluminense": ["Fluminense", "Fluminense-RJ", "Fluminense Football Club"],
  "Vasco": ["Vasco", "Vasco da Gama", "Vasco da Gama-RJ", "Club de Regatas Vasco da Gama"],
  "Botafogo": ["Botafogo", "Botafogo-RJ", "Botafogo RJ", "Botafogo de Futebol e Regatas"],
  "Palmeiras": ["Palmeiras", "Palmeiras-SP", "Sociedade Esportiva Palmeiras"],
  "Corinthians": ["Corinthians", "Corinthians-SP", "Sport Club Corinthians Paulista"],
  "São Paulo": ["São Paulo", "Sao Paulo", "São Paulo-SP", "Sao Paulo-SP",
    "São Paulo FC", "São Paulo Futebol Clube"],
  "Santos": ["Santos", "Santos-SP", "Santos FC", "Santos Futebol Clube"],
  "Grêmio": ["Grêmio", "Gremio", "Grêmio-RS", "Gremio-RS", "Grêmio Foot-Ball Porto Alegrense",
    "Grêmio - RS"],
  "Internacional": ["Internacional", "Internacional-RS", "Sport Club Internacional"],
  "Atlético-MG": ["Atlético-MG", "Atletico-MG", "Atlético Mineiro", "Atletico Mineiro",
    "Atlético - MG", "Atletico - MG", "Clube Atlético Mineiro"],
  "Cruzeiro": ["Cruzeiro", "Cruzeiro-MG", "Cruzeiro - MG", "Cruzeiro Esporte Clube"],
  "Athletico-PR": ["Athletico-PR", "Athletico Paranaense", "Atlético Paranaense", "Atlético-PR",
    "Atlético - PR", "Atletico Paranaense", "Atletico-PR", "Athletico Paranaense"],
  "Coritiba": ["Coritiba", "Coritiba-PR", "Coritiba - PR", "Coritiba Foot Ball Club"],
  "Bahia": ["Bahia", "Bahia-BA", "Bahia - BA", "EC Bahia", "Esporte Clube Bahia"],
  "Vitória": ["Vitória", "Vitoria", "Vitória-BA", "Vitória - BA", "Vitória EC",
    "Vitoria EC", "Esporte Clube Vitória"],
  "Ceará": ["Ceará", "Ceara", "Ceará - CE"],
  "Fortaleza": ["Fortaleza", "Fortaleza FC", "Fortaleza - CE",
    "Fortaleza Esporte Clube"],
  "Sport": ["Sport", "Sport-PE", "Sport Recife", "Sport-PE", "Sport Club do Recife"],
  "Náutico": ["Náutico", "Nautico", "Náutico-PE", "Nautico Capibaribe",
    "Clube Náutico Capibaribe"],
  "Goiás": ["Goiás", "Goias", "Goiás - GO", "Goiás Esporte Clube"],
  "Atlético-GO": ["Atlético-GO", "Atletico-GO", "Atlético - GO", "Atletico Goianiense",
    "Atlético Goianiense"],
  "Ponte Preta": ["Ponte Preta", "Ponte Preta-SP", "Associação Atlética Ponte Preta"],
  "Guarani": ["Guarani", "Guarani-SP", "Guarani SP", "Guarani Futebol Clube"],
  "Portuguesa": ["Portuguesa", "Portuguesa-SP", "Portuguesa RJ",
    "Associação Portuguesa de Desportos"],
  "Figueirense": ["Figueirense", "Figueirense-SC", "Figueirense - SC",
    "Figueirense Futebol Clube"],
  "Paraná": ["Paraná", "Parana", "Paraná - PR", "Paraná Clube"],
  "Juventude": ["Juventude", "Juventude-RS", "EC Juventude", "Esporte Clube Juventude"],
  "Criciúma": ["Criciúma", "Criciuma", "Criciuma - SC", "Criciúma Esporte Clube"],
  "Avaí": ["Avaí", "Avai", "Avaí - SC", "Avaí Futebol Clube"],
  "Joinville": ["Joinville", "Joinville - SC", "Joinville Esporte Clube"],
  "Chapecoense": ["Chapecoense", "Chapecoense - SC", "Associação Chapecoense de Futebol"],
  "Bragantino": ["Bragantino", "Bragantino - SP", "RB Bragantino",
    "Red Bull Bragantino", "Clube Atlético Bragantino"],
  "América-MG": ["América-MG", "America MG", "América - MG", "América Mineiro",
    "América Futebol Clube"],
  "São Caetano": ["São Caetano", "Sao Caetano", "Associação Desportiva São Caetano"],
  "Paysandu": ["Paysandu", "Paysandu - PA", "Paysandu Sport Club"],
  "Santa Cruz": ["Santa Cruz", "Santa Cruz FC", "Santa Cruz - PE",
    "Santa Cruz Futebol Clube"],
  "Cuiabá": ["Cuiabá", "Cuiaba", "Cuiabá - MT", "Cuiabá Esporte Clube"],
};

// Build reverse lookup map
for (const [canonical, variants] of Object.entries(BRAZILIAN_CLUBS)) {
  for (const variant of variants) {
    TEAM_NAME_MAP.set(normalizeKey(variant), canonical);
  }
}

function normalizeKey(name: string): string {
  return name
    .toLowerCase()
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "") // strip accents
    .replace(/[^a-z0-9]/g, "") // remove non-alphanumeric
    .trim();
}

/**
 * Normalize a team name to its canonical form.
 * Returns the canonical name if known, otherwise returns the cleaned input.
 */
export function normalizeTeam(name: string): string {
  if (!name) return "";
  const key = normalizeKey(name);
  return TEAM_NAME_MAP.get(key) || name.trim();
}

/**
 * Check if a team name (possibly with variations) matches a search query.
 * Both are normalized before comparison.
 */
export function teamMatches(teamName: string, query: string): boolean {
  const normalizedTeam = normalizeKey(normalizeTeam(teamName));
  const normalizedQuery = normalizeKey(query);
  return normalizedTeam.includes(normalizedQuery) || normalizedQuery.includes(normalizedTeam);
}

/**
 * Get the canonical name for display.
 */
export function getCanonicalName(name: string): string {
  return normalizeTeam(name);
}

/**
 * Get all known team names.
 */
export function getAllTeamNames(): string[] {
  return Array.from(new Set(TEAM_NAME_MAP.values())).sort();
}

File: experiments-local/experiment-17/runs/agent=omp_language=csharp_model=openrouter/deepseek/deepseek-v4-pro_tooling=none/rep1/BrazilianSoccerMcp/Data/TeamNormalizer.cs
using System.Text.RegularExpressions;

namespace BrazilianSoccerMcp.Data;

/// <summary>
/// Normalises Brazilian team names across datasets so "Flamengo-RJ",
/// "Flamengo", and "Clube de Regatas do Flamengo" all resolve to the
/// same canonical key.
/// </summary>
public static partial class TeamNormalizer
{
    /// <summary>Map from canonical key to display name.</summary>
    private static readonly Dictionary<string, string> CanonicalToDisplay = new(StringComparer.OrdinalIgnoreCase)
    {
        ["flamengo"] = "Flamengo",
        ["fluminense"] = "Fluminense",
        ["vasco"] = "Vasco da Gama",
        ["botafogo"] = "Botafogo",
        ["palmeiras"] = "Palmeiras",
        ["corinthians"] = "Corinthians",
        ["sao paulo"] = "São Paulo",
        ["santos"] = "Santos",
        ["gremio"] = "Grêmio",
        ["internacional"] = "Internacional",
        ["atletico mg"] = "Atlético-MG",
        ["atletico mineiro"] = "Atlético-MG",
        ["atletico pr"] = "Athletico-PR",
        ["athletico pr"] = "Athletico-PR",
        ["athletico paranaense"] = "Athletico-PR",
        ["cruzeiro"] = "Cruzeiro",
        ["bahia"] = "Bahia",
        ["sport"] = "Sport",
        ["sport recife"] = "Sport",
        ["fortaleza"] = "Fortaleza",
        ["ceara"] = "Ceará",
        ["goias"] = "Goiás",
        ["coritiba"] = "Coritiba",
        ["chapecoense"] = "Chapecoense",
        ["avai"] = "Avaí",
        ["vitoria"] = "Vitória",
        ["ponte preta"] = "Ponte Preta",
        ["figueirense"] = "Figueirense",
        ["nautico"] = "Náutico",
        ["parana"] = "Paraná",
        ["juventude"] = "Juventude",
        ["criciuma"] = "Criciúma",
        ["joinville"] = "Joinville",
        ["santa cruz"] = "Santa Cruz",
        ["america mg"] = "América-MG",
        ["america mineiro"] = "América-MG",
        ["america rn"] = "América-RN",
        ["bragantino"] = "Bragantino",
        ["red bull bragantino"] = "Bragantino",
        ["guarani"] = "Guarani",
        ["portuguesa"] = "Portuguesa",
        ["paysandu"] = "Paysandu",
        ["gama"] = "Gama",
        ["brasiliense"] = "Brasiliense",
        ["sao caetano"] = "São Caetano",
        ["boca juniors"] = "Boca Juniors",
        ["river plate"] = "River Plate",
        ["racing"] = "Racing",
        ["independiente"] = "Independiente",
        ["san lorenzo"] = "San Lorenzo",
        ["estudiantes"] = "Estudiantes",
        ["velez sarsfield"] = "Vélez Sarsfield",
        ["penarol"] = "Peñarol",
        ["nacional"] = "Nacional",
        ["olimpia"] = "Olimpia",
        ["cerro porteno"] = "Cerro Porteño",
        ["libertad"] = "Libertad",
        ["colo colo"] = "Colo-Colo",
        ["universidad de chile"] = "Universidad de Chile",
        ["universidad catolica"] = "Universidad Católica",
        ["barcelona"] = "Barcelona-EQU",
        ["barcelona equ"] = "Barcelona-EQU",
        ["emelec"] = "Emelec",
        ["liga de quito"] = "Liga de Quito",
        ["sporting cristal"] = "Sporting Cristal",
        ["alianza lima"] = "Alianza Lima",
        ["universitario"] = "Universitario",
        ["bolivar"] = "Bolívar",
        ["the strongest"] = "The Strongest",
        ["toluca"] = "Toluca",
        ["tigres"] = "Tigres",
        ["club america"] = "América-MEX",
        ["cruz azul"] = "Cruz Azul",
        ["atletico nacional"] = "Atlético Nacional",
        ["junior"] = "Junior",
        ["independiente medellin"] = "Independiente Medellín",
        ["millonarios"] = "Millonarios",
        ["santa fe"] = "Santa Fe",
        ["deportivo cali"] = "Deportivo Cali",
        ["caracas"] = "Caracas",
        ["deportivo tachira"] = "Deportivo Táchira",
    };

    // Patterns to strip from team names
    [GeneratedRegex(@"\s*-\s*(SP|RJ|MG|RS|PR|SC|BA|PE|CE|GO|DF|RN|AL|PA|AM|MA|PB|SE|ES|MT|MS|PI|RO|TO|AP|AC|RR)$", RegexOptions.IgnoreCase)]
    private static partial Regex StateSuffixPattern();

    [GeneratedRegex(@"\(.*?\)")]
    private static partial Regex ParenthesesPattern();

    [GeneratedRegex(@"\s+")]
    private static partial Regex WhitespacePattern();

    /// <summary>
    /// Returns the canonical key for a raw team name. Strips state suffixes,
    /// parenthesised qualifiers, </summary>
    public static string Normalize(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw))
            return string.Empty;

        var name = raw.Trim();

        // Strip common long-form prefixes/suffixes
        name = StripClubPrefixes(name);

        // Strip state suffix: "Flamengo-RJ" → "Flamengo"
        name = StateSuffixPattern().Replace(name, "");

        // Strip parenthesised qualifiers: "Nacional (URU)" → "Nacional"
        name = ParenthesesPattern().Replace(name, "");

        // Strip " - UF" style (common in Brazilian Cup)
        // e.g., "Gama - DF" → "Gama"
        name = Regex.Replace(name, @"\s*-\s*(SP|RJ|MG|RS|PR|SC|BA|PE|CE|GO|DF|RN|AL|PA|AM|MA|PB|SE|ES|MT|MS|PI|RO|TO|AP|AC|RR)$", "", RegexOptions.IgnoreCase);

        // Normalize whitespace and case
        name = WhitespacePattern().Replace(name, " ").Trim().ToLowerInvariant()
            .Replace("á", "a").Replace("à", "a").Replace("ã", "a").Replace("â", "a")
            .Replace("é", "e").Replace("ê", "e")
            .Replace("í", "i")
            .Replace("ó", "o").Replace("ô", "o").Replace("õ", "o")
            .Replace("ú", "u").Replace("ü", "u")
            .Replace("ç", "c");

        // Try direct lookup in canonical map
        if (CanonicalToDisplay.TryGetValue(name, out var display))
            return display;

        // Try matching without accent-stripped version
        foreach (var kvp in CanonicalToDisplay)
        {
            if (kvp.Key == name)
                return kvp.Value;
        }

        // Return title-cased raw name as fallback
        return CultureAwareTitleCase(name);
    }

    private static string StripClubPrefixes(string name)
    {
        // Strip common prefixes
        var prefixes = new[]
        {
            "sport club", "clube de regatas", "clube atletico", "atletico clube",
            "associacao atletica", "sociedade esportiva", "esporte clube",
            "gremio esportivo", "gremio foot-ball", "foot ball club",
            "sport club do", "clube do", "sport club", "club de regatas",
            "botafogo de futebol e regatas", "sao paulo futebol clube",
            "santos futebol clube",
        };

        var lower = name.ToLowerInvariant().Trim();
        foreach (var prefix in prefixes)
        {
            if (lower.StartsWith(prefix + " "))
                return name[(prefix.Length + 1)..].Trim();
        }
        return name;
    }

    private static string CultureAwareTitleCase(string name)
    {
        // Simple title casing preserving Portuguese conventions
        var words = name.Split(' ', StringSplitOptions.RemoveEmptyEntries);
        for (int i = 0; i < words.Length; i++)
        {
            if (words[i].Length > 0)
            {
                var prepositions = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
                { "de", "da", "do", "das", "dos", "e", "a", "o", "no", "na", "em" };
                if (i > 0 && i < words.Length - 1 && prepositions.Contains(words[i]))
                    words[i] = words[i].ToLowerInvariant();
                else
                    words[i] = char.ToUpperInvariant(words[i][0]) + words[i][1..].ToLowerInvariant();
            }
        }
        return string.Join(" ", words);
    }

    /// <summary>Check if the normalised name contains a search term.</summary>
    public static bool Matches(string? teamName, string searchTerm)
    {
        var normalized = Normalize(teamName);
        var search = searchTerm.ToLowerInvariant()
            .Replace("á", "a").Replace("ã", "a").Replace("â", "a")
            .Replace("é", "e").Replace("ê", "e")
            .Replace("í", "i")
            .Replace("ó", "o").Replace("ô", "o").Replace("õ", "o")
            .Replace("ú", "u").Replace("ç", "c");
        return normalized.ToLowerInvariant().Contains(search);
    }
}